File: packages/server/src/features/orchestration/orchestration.service.ts
import { prisma } from '../../lib/prisma.js';
import { CursorApiClient } from '../../lib/cursor-api.js';
import { broadcastToOrchestration } from '../websocket/websocket.service.js';
import { 
  PlanningAgentOutputSchema, 
  PLANNING_AGENT_SYSTEM_PROMPT,
  buildPlanningAgentPrompt,
  type PlanningAgentOutput 
} from './orchestration.prompts.js';
import { ApiError } from '../../lib/validation.js';

export class OrchestrationService {
  private cursorClient: CursorApiClient;

  constructor(cursorApiKey: string) {
    this.cursorClient = new CursorApiClient(cursorApiKey);
  }

  async createOrchestration(params: {
    userId: string;
    repository: string;
    ref?: string;
    initialPrompt: string;
  }) {
    const orchestration = await prisma.orchestration.create({
      data: {
        userId: params.userId,
        repository: params.repository,
        ref: params.ref || 'main',
        initialPrompt: params.initialPrompt,
        status: 'PLANNING'
      }
    });

    await this.createEvent(orchestration.id, 'orchestration_created', {
      repository: params.repository,
      prompt: params.initialPrompt
    });

    this.startPlanningPhase(orchestration.id).catch(err => {
      console.error('Error in planning phase:', err);
      this.handleOrchestrationError(orchestration.id, err);
    });

    return orchestration;
  }

  private async startPlanningPhase(orchestrationId: string) {
    const orchestration = await prisma.orchestration.findUnique({
      where: { id: orchestrationId },
      include: { user: true, followUpMessages: true }
    });

    if (!orchestration) {
      throw new ApiError(404, 'Orchestration not found');
    }

    const previousQA = orchestration.followUpMessages
      .sort((a, b) => a.createdAt.getTime() - b.createdAt.getTime())
      .reduce((acc, msg, idx, arr) => {
        if (msg.role === 'agent' && arr[idx + 1]?.role === 'user') {
          acc.push({
            question: msg.content,
            answer: arr[idx + 1].content
          });
        }
        return acc;
      }, [] as Array<{ question: string; answer: string }>);

    const prompt = buildPlanningAgentPrompt(
      orchestration.initialPrompt,
      orchestration.repository,
      previousQA.length > 0 ? previousQA : undefined
    );

    const fullPrompt = `${PLANNING_AGENT_SYSTEM_PROMPT}\n\n${prompt}`;

    const cursorAgent = await this.cursorClient.createAgent({
      prompt: { text: fullPrompt },
      source: {
        repository: orchestration.repository,
        ref: orchestration.ref
      },
      target: {
        branchName: `planning/${orchestrationId}`,
        autoCreatePr: false
      }
    });

    await this.createEvent(orchestrationId, 'planning_agent_created', {
      cursorAgentId: cursorAgent.id
    });

    this.pollPlanningAgent(orchestrationId, cursorAgent.id).catch(err => {
      console.error('Error polling planning agent:', err);
      this.handleOrchestrationError(orchestrationId, err);
    });
  }

  private async pollPlanningAgent(orchestrationId: string, cursorAgentId: string) {
    let attempts = 0;
    const maxAttempts = 120;
    const pollInterval = 5000;

    while (attempts < maxAttempts) {
      await new Promise(resolve => setTimeout(resolve, pollInterval));
      attempts++;

      try {
        const agent = await this.cursorClient.getAgent(cursorAgentId);
        
        if (agent.status === 'COMPLETED' || agent.status === 'completed') {
          const conversation = await this.cursorClient.getAgentConversation(cursorAgentId);
          await this.processPlanningOutput(orchestrationId, conversation);
          break;
        } else if (agent.status === 'FAILED' || agent.status === 'failed') {
          throw new Error('Planning agent failed');
        }
      } catch (error) {
        console.error('Error polling agent:', error);
        throw error;
      }
    }

    if (attempts >= maxAttempts) {
      throw new Error('Planning agent timeout');
    }
  }

  private async processPlanningOutput(orchestrationId: string, conversation: any) {
    const lastMessage = conversation.messages
      .filter((m: any) => m.type === 'assistant_message')
      .pop();

    if (!lastMessage) {
      throw new Error('No output from planning agent');
    }

    let output: PlanningAgentOutput;
    try {
      const jsonMatch = lastMessage.text.match(/\{[\s\S]*\}/);
      if (!jsonMatch) {
        throw new Error('No JSON found in agent output');
      }
      const parsed = JSON.parse(jsonMatch[0]);
      output = PlanningAgentOutputSchema.parse(parsed);
    } catch (error) {
      console.error('Failed to parse planning output:', error);
      console.error('Agent output:', lastMessage.text);
      throw new Error('Invalid planning agent output format');
    }

    if (output.type === 'questions') {
      await prisma.orchestration.update({
        where: { id: orchestrationId },
        data: { 
          status: 'AWAITING_FOLLOWUP',
          planningOutput: output as any
        }
      });

      const questions = 'questions' in output.content ? output.content.questions : [];
      for (const q of questions) {
        await prisma.followUpMessage.create({
          data: {
            orchestrationId,
            role: 'agent',
            content: q.question
          }
        });
      }

      await this.createEvent(orchestrationId, 'questions_asked', {
        questions
      });

      broadcastToOrchestration(orchestrationId, {
        type: 'questions_asked',
        questions
      });
    } else {
      await prisma.orchestration.update({
        where: { id: orchestrationId },
        data: { 
          status: 'AWAITING_APPROVAL',
          planningOutput: output as any
        }
      });

      await this.createEvent(orchestrationId, 'plan_ready', {
        plan: output.content
      });

      broadcastToOrchestration(orchestrationId, {
        type: 'plan_ready',
        plan: output.content
      });
    }
  }

  async answerFollowUpQuestions(
    orchestrationId: string,
    answers: Record<string, string>
  ) {
    const orchestration = await prisma.orchestration.findUnique({
      where: { id: orchestrationId }
    });

    if (!orchestration || orchestration.status !== 'AWAITING_FOLLOWUP') {
      throw new ApiError(400, 'Invalid orchestration state');
    }

    for (const [questionId, answer] of Object.entries(answers)) {
      await prisma.followUpMessage.create({
        data: {
          orchestrationId,
          role: 'user',
          content: answer
        }
      });
    }

    await prisma.orchestration.update({
      where: { id: orchestrationId },
      data: { status: 'PLANNING' }
    });

    await this.createEvent(orchestrationId, 'followup_answered', { answers });

    this.startPlanningPhase(orchestrationId).catch(err => {
      console.error('Error restarting planning:', err);
      this.handleOrchestrationError(orchestrationId, err);
    });
  }

  async approvePlan(orchestrationId: string) {
    const orchestration = await prisma.orchestration.findUnique({
      where: { id: orchestrationId },
      include: { user: true }
    });

    if (!orchestration || orchestration.status !== 'AWAITING_APPROVAL') {
      throw new ApiError(400, 'Invalid orchestration state');
    }

    const plan = orchestration.planningOutput as any;

    await prisma.orchestration.update({
      where: { id: orchestrationId },
      data: { 
        status: 'EXECUTING',
        approvedPlan: plan
      }
    });

    await this.createEvent(orchestrationId, 'plan_approved', {});

    this.executePlan(orchestrationId, orchestration.user.cursorApiKey, plan.content).catch(err => {
      console.error('Error executing plan:', err);
      this.handleOrchestrationError(orchestrationId, err);
    });
  }

  private async executePlan(
    orchestrationId: string,
    cursorApiKey: string,
    plan: any
  ) {
    if (plan.requiresSubAgents && plan.subAgents) {
      for (const subAgentSpec of plan.subAgents) {
        await this.createSubAgent(orchestrationId, cursorApiKey, subAgentSpec);
      }
    } else {
      await this.createMainExecutionAgent(orchestrationId, cursorApiKey, plan);
    }

    this.monitorExecution(orchestrationId).catch(err => {
      console.error('Error monitoring execution:', err);
    });
  }

  private async createSubAgent(
    orchestrationId: string,
    cursorApiKey: string,
    spec: any
  ) {
    const orchestration = await prisma.orchestration.findUnique({
      where: { id: orchestrationId }
    });

    if (!orchestration) return;

    const client = new CursorApiClient(cursorApiKey);

    const cursorAgent = await client.createAgent({
      prompt: { text: spec.prompt },
      source: {
        repository: orchestration.repository,
        ref: orchestration.ref
      },
      target: {
        branchName: spec.branchName || `feature/${spec.id}`,
        autoCreatePr: true
      }
    });

    await prisma.agent.create({
      data: {
        orchestrationId,
        cursorAgentId: cursorAgent.id,
        name: spec.name,
        prompt: spec.prompt,
        status: 'CREATING',
        branchName: spec.branchName,
        metadata: { tasks: spec.tasks }
      }
    });

    await this.createEvent(orchestrationId, 'agent_spawned', {
      agentId: cursorAgent.id,
      name: spec.name
    });

    broadcastToOrchestration(orchestrationId, {
      type: 'agent_spawned',
      agent: {
        id: cursorAgent.id,
        name: spec.name,
        status: 'CREATING'
      }
    });
  }

  private async createMainExecutionAgent(
    orchestrationId: string,
    cursorApiKey: string,
    plan: any
  ) {
    const orchestration = await prisma.orchestration.findUnique({
      where: { id: orchestrationId }
    });

    if (!orchestration) return;

    const client = new CursorApiClient(cursorApiKey);

    const prompt = `${orchestration.initialPrompt}\n\nExecution Plan:\n${JSON.stringify(plan.tasks, null, 2)}`;

    const cursorAgent = await client.createAgent({
      prompt: { text: prompt },
      source: {
        repository: orchestration.repository,
        ref: orchestration.ref
      },
      target: {
        branchName: `feature/${orchestrationId}`,
        autoCreatePr: true
      }
    });

    await prisma.agent.create({
      data: {
        orchestrationId,
        cursorAgentId: cursorAgent.id,
        name: 'Main Execution Agent',
        prompt,
        status: 'CREATING',
        branchName: `feature/${orchestrationId}`,
        metadata: { tasks: plan.tasks }
      }
    });

    await this.createEvent(orchestrationId, 'agent_spawned', {
      agentId: cursorAgent.id,
      name: 'Main Execution Agent'
    });
  }

  private async monitorExecution(orchestrationId: string) {
    const pollInterval = 10000;
    
    while (true) {
      await new Promise(resolve => setTimeout(resolve, pollInterval));

      const agents = await prisma.agent.findMany({
        where: { orchestrationId }
      });

      if (agents.length === 0) break;

      let allCompleted = true;
      let anyFailed = false;

      for (const agent of agents) {
        if (!agent.cursorAgentId) continue;

        try {
          const cursorAgent = await this.cursorClient.getAgent(agent.cursorAgentId);
          
          const newStatus = this.mapCursorStatus(cursorAgent.status);
          
          if (agent.status !== newStatus) {
            await prisma.agent.update({
              where: { id: agent.id },
              data: { 
                status: newStatus,
                pullRequestUrl: cursorAgent.target?.url
              }
            });

            await prisma.agentStatusUpdate.create({
              data: {
                agentId: agent.id,
                status: newStatus,
                message: `Status changed to ${newStatus}`
              }
            });

            broadcastToOrchestration(orchestrationId, {
              type: 'agent_status_update',
              agent: {
                id: agent.id,
                cursorAgentId: agent.cursorAgentId,
                name: agent.name,
                status: newStatus,
                pullRequestUrl: cursorAgent.target?.url
              }
            });
          }

          if (newStatus !== 'COMPLETED') allCompleted = false;
          if (newStatus === 'FAILED') anyFailed = true;
        } catch (error) {
          console.error(`Error checking agent ${agent.id}:`, error);
        }
      }

      if (allCompleted || anyFailed) {
        const finalStatus = anyFailed ? 'FAILED' : 'COMPLETED';
        await prisma.orchestration.update({
          where: { id: orchestrationId },
          data: { 
            status: finalStatus,
            completedAt: new Date()
          }
        });

        await this.createEvent(orchestrationId, 'orchestration_completed', {
          status: finalStatus
        });

        broadcastToOrchestration(orchestrationId, {
          type: 'orchestration_completed',
          status: finalStatus
        });

        break;
      }
    }
  }

  private mapCursorStatus(status: string): any {
    const statusMap: Record<string, string> = {
      'CREATING': 'CREATING',
      'RUNNING': 'RUNNING',
      'COMPLETED': 'COMPLETED',
      'completed': 'COMPLETED',
      'FAILED': 'FAILED',
      'failed': 'FAILED',
      'CANCELLED': 'CANCELLED',
      'cancelled': 'CANCELLED'
    };
    return statusMap[status] || 'RUNNING';
  }

  private async createEvent(orchestrationId: string, type: string, data: any) {
    await prisma.orchestrationEvent.create({
      data: {
        orchestrationId,
        type,
        data
      }
    });
  }

  private async handleOrchestrationError(orchestrationId: string, error: any) {
    await prisma.orchestration.update({
      where: { id: orchestrationId },
      data: { status: 'FAILED' }
    });

    await this.createEvent(orchestrationId, 'error', {
      message: error.message,
      stack: error.stack
    });

    broadcastToOrchestration(orchestrationId, {
      type: 'error',
      message: error.message
    });
  }
}

File: packages/dotnet-api/Source/Features/Orchestration/Services/OrchestrationService.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using Microsoft.AspNetCore.SignalR;
using Microsoft.EntityFrameworkCore;
using Source.Infrastructure;
using Source.Infrastructure.Services.CursorApi;
using Source.Features.Orchestration.Hubs;
using Hangfire;
using OrchestrationModel = Source.Features.Orchestration.Models.Orchestration;
using AgentModel = Source.Features.Orchestration.Models.Agent;
using OrchestrationStatusEnum = Source.Features.Orchestration.Models.OrchestrationStatus;
using AgentStatusEnum = Source.Features.Orchestration.Models.AgentStatus;
using FollowUpMessageModel = Source.Features.Orchestration.Models.FollowUpMessage;
using OrchestrationEventModel = Source.Features.Orchestration.Models.OrchestrationEvent;

namespace Source.Features.Orchestration.Services;

public class OrchestrationService
{
    private readonly ApplicationDbContext _context;
    private readonly IHubContext<OrchestrationHub> _hubContext;
    private readonly ILogger<OrchestrationService> _logger;

    public OrchestrationService(
        ApplicationDbContext context,
        IHubContext<OrchestrationHub> hubContext,
        ILogger<OrchestrationService> logger)
    {
        _context = context;
        _hubContext = hubContext;
        _logger = logger;
    }

    public async Task<OrchestrationModel> CreateOrchestrationAsync(
        string userId,
        string cursorApiKey,
        string repository,
        string initialPrompt,
        string? refBranch = null)
    {
        var orchestration = new OrchestrationModel
        {
            UserId = userId,
            Repository = repository,
            Ref = refBranch ?? "main",
            InitialPrompt = initialPrompt,
            Status = OrchestrationStatusEnum.PLANNING
        };

        _context.Orchestrations.Add(orchestration);
        await _context.SaveChangesAsync();

        await CreateEventAsync(orchestration.Id, "orchestration_created", new
        {
            repository,
            prompt = initialPrompt
        });

        BackgroundJob.Enqueue<OrchestrationBackgroundJobs>(x =>
            x.StartPlanningPhaseAsync(orchestration.Id, cursorApiKey));

        return orchestration;
    }

    public async Task<OrchestrationModel?> GetOrchestrationAsync(string orchestrationId, string userId)
    {
        return await _context.Orchestrations
            .Include(o => o.Agents)
                .ThenInclude(a => a.StatusUpdates.OrderByDescending(s => s.CreatedAt).Take(5))
            .Include(o => o.FollowUpMessages.OrderBy(f => f.CreatedAt))
            .Include(o => o.Events.OrderByDescending(e => e.CreatedAt).Take(20))
            .FirstOrDefaultAsync(o => o.Id == orchestrationId && o.UserId == userId);
    }

    public async Task<List<OrchestrationModel>> ListOrchestrationsAsync(string userId)
    {
        return await _context.Orchestrations
            .Include(o => o.Agents)
            .Where(o => o.UserId == userId)
            .OrderByDescending(o => o.CreatedAt)
            .ToListAsync();
    }

    public async Task StartPlanningPhaseAsync(string orchestrationId, string cursorApiKey)
    {
        var orchestration = await _context.Orchestrations
            .Include(o => o.FollowUpMessages.OrderBy(f => f.CreatedAt))
            .FirstOrDefaultAsync(o => o.Id == orchestrationId);

        if (orchestration == null)
        {
            _logger.LogError("Orchestration {OrchestrationId} not found", orchestrationId);
            return;
        }

        var previousQA = orchestration.FollowUpMessages
            .Where(m => m.Role == "agent")
            .Select((agentMsg, index) =>
            {
                var nextUserMsg = orchestration.FollowUpMessages
                    .FirstOrDefault(m => m.Role == "user" && m.CreatedAt > agentMsg.CreatedAt);
                return nextUserMsg != null
                    ? (question: agentMsg.Content, answer: nextUserMsg.Content)
                    : ((string, string)?)null;
            })
            .Where(qa => qa.HasValue)
            .Select(qa => qa!.Value)
            .ToList();

        var prompt = OrchestrationPrompts.BuildPlanningAgentPrompt(
            orchestration.InitialPrompt,
            orchestration.Repository,
            previousQA.Count > 0 ? previousQA : null);

        var fullPrompt = $"{OrchestrationPrompts.PLANNING_AGENT_SYSTEM_PROMPT}\n\n{prompt}";

        var httpClient = new HttpClient();
        var cursorClient = new CursorApiClient(cursorApiKey, httpClient);

        var cursorAgent = await cursorClient.CreateAgentAsync(new CursorAgentCreateRequest
        {
            Prompt = new CursorAgentCreateRequest.PromptData { Text = fullPrompt },
            Source = new CursorAgentCreateRequest.SourceData
            {
                Repository = orchestration.Repository,
                Ref = orchestration.Ref
            },
            Target = new CursorAgentCreateRequest.TargetData
            {
                BranchName = $"planning/{orchestrationId}",
                AutoCreatePr = false
            }
        });

        orchestration.PlanningAgentId = cursorAgent.Id;
        await _context.SaveChangesAsync();

        await CreateEventAsync(orchestrationId, "planning_agent_created", new
        {
            cursorAgentId = cursorAgent.Id
        });

        BackgroundJob.Enqueue<OrchestrationBackgroundJobs>(x =>
            x.PollPlanningAgentAsync(orchestrationId, cursorAgent.Id, cursorApiKey));
    }

    public async Task ProcessPlanningOutputAsync(string orchestrationId, string conversationJson)
    {
        var orchestration = await _context.Orchestrations.FindAsync(orchestrationId);
        if (orchestration == null) return;

        var conversation = JsonSerializer.Deserialize<CursorAgentConversation>(conversationJson);
        var lastMessage = conversation?.Messages
            .Where(m => m.Type == "assistant_message")
            .LastOrDefault();

        if (lastMessage == null)
        {
            await HandleOrchestrationErrorAsync(orchestrationId, "No output from planning agent");
            return;
        }

        try
        {
            var jsonMatch = System.Text.RegularExpressions.Regex.Match(lastMessage.Text, @"\{[\s\S]*\}");
            if (!jsonMatch.Success)
            {
                await HandleOrchestrationErrorAsync(orchestrationId, "No JSON found in agent output");
                return;
            }

            var outputJson = jsonMatch.Value;
            var output = JsonNode.Parse(outputJson);
            var type = output?["type"]?.GetValue<string>();

            if (type == "questions")
            {
                orchestration.Status = OrchestrationStatusEnum.AWAITING_FOLLOWUP;
                orchestration.PlanningOutput = outputJson;
                await _context.SaveChangesAsync();

                var questions = output["content"]?["questions"]?.AsArray();
                if (questions != null)
                {
                    foreach (var q in questions)
                    {
                        var question = q?["question"]?.GetValue<string>();
                        if (!string.IsNullOrEmpty(question))
                        {
                            _context.FollowUpMessages.Add(new FollowUpMessageModel
                            {
                                OrchestrationId = orchestrationId,
                                Role = "agent",
                                Content = question
                            });
                        }
                    }
                    await _context.SaveChangesAsync();
                }

                await CreateEventAsync(orchestrationId, "questions_asked", new { questions });
                await BroadcastToOrchestrationAsync(orchestrationId, new { type = "questions_asked", questions });
            }
            else if (type == "plan")
            {
                orchestration.Status = OrchestrationStatusEnum.AWAITING_APPROVAL;
                orchestration.PlanningOutput = outputJson;
                await _context.SaveChangesAsync();

                var plan = output["content"];
                await CreateEventAsync(orchestrationId, "plan_ready", new { plan });
                await BroadcastToOrchestrationAsync(orchestrationId, new { type = "plan_ready", plan });
            }
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to parse planning output for {OrchestrationId}", orchestrationId);
            await HandleOrchestrationErrorAsync(orchestrationId, $"Invalid planning output: {ex.Message}");
        }
    }

    public async Task AnswerFollowUpQuestionsAsync(
        string orchestrationId,
        string cursorApiKey,
        Dictionary<string, string> answers)
    {
        var orchestration = await _context.Orchestrations.FindAsync(orchestrationId);
        if (orchestration == null || orchestration.Status != OrchestrationStatusEnum.AWAITING_FOLLOWUP)
        {
            throw new InvalidOperationException("Invalid orchestration state");
        }

        if (string.IsNullOrEmpty(orchestration.PlanningAgentId))
        {
            throw new InvalidOperationException("No planning agent found");
        }

        foreach (var answer in answers.Values)
        {
            _context.FollowUpMessages.Add(new FollowUpMessageModel
            {
                OrchestrationId = orchestrationId,
                Role = "user",
                Content = answer
            });
        }

        orchestration.Status = OrchestrationStatusEnum.PLANNING;
        await _context.SaveChangesAsync();

        var answersText = string.Join("\n\n", answers.Values);
        var followUpText = $"Here are the answers to your questions:\n\n{answersText}\n\nPlease continue with the planning.";

        var httpClient = new HttpClient();
        var cursorClient = new CursorApiClient(cursorApiKey, httpClient);

        await cursorClient.AddFollowUpAsync(
            orchestration.PlanningAgentId,
            new CursorAgentCreateRequest.PromptData { Text = followUpText });

        await CreateEventAsync(orchestrationId, "followup_answered", new { answers });

        BackgroundJob.Enqueue<OrchestrationBackgroundJobs>(x =>
            x.PollPlanningAgentAsync(orchestrationId, orchestration.PlanningAgentId, cursorApiKey));
    }

    public async Task ApprovePlanAsync(string orchestrationId, string cursorApiKey)
    {
        var orchestration = await _context.Orchestrations.FindAsync(orchestrationId);
        if (orchestration == null || orchestration.Status != OrchestrationStatusEnum.AWAITING_APPROVAL)
        {
            throw new InvalidOperationException("Invalid orchestration state");
        }

        orchestration.Status = OrchestrationStatusEnum.EXECUTING;
        orchestration.ApprovedPlan = orchestration.PlanningOutput;
        await _context.SaveChangesAsync();

        await CreateEventAsync(orchestrationId, "plan_approved", new { });

        BackgroundJob.Enqueue<OrchestrationBackgroundJobs>(x =>
            x.ExecutePlanAsync(orchestrationId, cursorApiKey));
    }

    public async Task CancelOrchestrationAsync(string orchestrationId, string cursorApiKey)
    {
        var orchestration = await _context.Orchestrations
            .Include(o => o.Agents)
            .FirstOrDefaultAsync(o => o.Id == orchestrationId);

        if (orchestration == null) return;

        orchestration.Status = OrchestrationStatusEnum.CANCELLED;
        await _context.SaveChangesAsync();

        var httpClient = new HttpClient();
        var cursorClient = new CursorApiClient(cursorApiKey, httpClient);

        foreach (var agent in orchestration.Agents.Where(a => a.Status == AgentStatusEnum.RUNNING))
        {
            try
            {
                if (!string.IsNullOrEmpty(agent.CursorAgentId))
                {
                    await cursorClient.CancelAgentAsync(agent.CursorAgentId);
                }
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to cancel agent {AgentId}", agent.Id);
            }
        }
    }

    public async Task CreateEventAsync(string orchestrationId, string type, object data)
    {
        var dataJson = JsonSerializer.Serialize(data);
        _context.OrchestrationEvents.Add(new OrchestrationEventModel
        {
            OrchestrationId = orchestrationId,
            Type = type,
            Data = dataJson
        });
        await _context.SaveChangesAsync();
    }

    public async Task HandleOrchestrationErrorAsync(string orchestrationId, string errorMessage)
    {
        var orchestration = await _context.Orchestrations.FindAsync(orchestrationId);
        if (orchestration != null)
        {
            orchestration.Status = OrchestrationStatusEnum.FAILED;
            await _context.SaveChangesAsync();
        }

        await CreateEventAsync(orchestrationId, "error", new { message = errorMessage });
        await BroadcastToOrchestrationAsync(orchestrationId, new { type = "error", message = errorMessage });
    }

    public async Task BroadcastToOrchestrationAsync(string orchestrationId, object message)
    {
        await _hubContext.Clients
            .Group(orchestrationId)
            .SendAsync("BroadcastToOrchestration", message);
    }
}

